zrml/prediction-markets/src/migrations.rs

```diff
@@ -434,6 +434,67 @@ mod tests {
         });
     }
 
+    #[test]
+    fn test_dispute_blocks_with_gap_of_dispute_duration() {
+        ExtBuilder::default().build().execute_with(|| {
+            setup_chain();
+
+            System::set_block_number(1);
+            let market_ids_reported_0 =
+                BoundedVec::<MarketIdOf<Runtime>, CacheSize>::try_from(vec![0])
+                    .expect("boundedvec creation failed");
+            let market_ids_disputed_0 =
+                BoundedVec::<MarketIdOf<Runtime>, CacheSize>::try_from(vec![1])
+                    .expect("boundedvec creation failed");
+            System::set_block_number(4);
+            let dispute_block = System::current_block_number().saturating_sub(1_u32.into());
+            MarketIdsPerDisputeBlock::<Runtime>::insert(
+                dispute_block,
+                market_ids_disputed_0.clone(),
+            );
+            MarketIdsPerReportBlock::<Runtime>::insert(
+                dispute_block,
+                market_ids_reported_0.clone(),
+            );
+            let dispute_duration = <Runtime as Config>::DisputePeriod::get();
+            let next_dispute_block = dispute_block + dispute_duration;
+            let market_ids_reported_1 =
+                BoundedVec::<MarketIdOf<Runtime>, CacheSize>::try_from(vec![2])
+                    .expect("boundedvec creation failed");
+            let market_ids_disputed_1 =
+                BoundedVec::<MarketIdOf<Runtime>, CacheSize>::try_from(vec![3])
+                    .expect("boundedvec creation failed");
+            MarketIdsPerDisputeBlock::<Runtime>::insert(
+                next_dispute_block,
+                market_ids_disputed_1.clone(),
+            );
+            MarketIdsPerReportBlock::<Runtime>::insert(
+                next_dispute_block,
+                market_ids_reported_1.clone(),
+            );
+            MigrateMarketIdsPerBlockStorage::<Runtime>::on_runtime_upgrade();
+            // after migration dispute_block's data is stored against next_dispute_block
+            assert_eq!(
+                MarketIdsPerDisputeBlock::<Runtime>::get(next_dispute_block),
+                market_ids_disputed_0
+            );
+            assert_eq!(
+                MarketIdsPerReportBlock::<Runtime>::get(next_dispute_block),
+                market_ids_reported_0
+            );
+            // after migration next_dispute_block's data is stored against
+            // next_dispute_block + dispute_duration
+            assert_eq!(
+                MarketIdsPerDisputeBlock::<Runtime>::get(next_dispute_block + dispute_duration),
+                market_ids_disputed_1
+            );
+            assert_eq!(
+                MarketIdsPerReportBlock::<Runtime>::get(next_dispute_block + dispute_duration),
+                market_ids_reported_1
+            );
+        });
+    }
```

Comment:
Much better with this test. Your solution [here](https://github.com/zeitgeistpm/zeitgeist/blob/12e3cc8ffd0d88d9f69d3f372c1bf7ccad0d8bf7/zrml/prediction-markets/src/migrations.rs#L204-L212) solved the problem. First I thought you just added this test, but did not make the mentioned change and then I debugged it. I was ultra confused as I saw, that the for loop started with the last element in my test in the following picture.

<img width="990" alt="Bildschirmfoto 2022-10-03 um 11 18 03" src="https://user-images.githubusercontent.com/33418719/193544051-d0c3d37d-29dd-4865-8ad7-48fbed1c2b82.png">

Then I realised, that the order of iteration is not guaranteed in `StorageMap` iterations. So yeah, non-deterministic problems 🤯. 

> Enumerate all keys in the map in no particular order.

(https://paritytech.github.io/substrate/master/frame_support/pallet_prelude/struct.StorageMap.html#method.iter_keys)